### PR TITLE
Spatial_Engine: Remove Normal from RoundCoordinates

### LINKED ISSUE #2441: Spatial_Engine: Couldn't RoundCoordinates(IElement2D) extract the normal based on fit plane?
#### Description:
As in the title, couldn't the method call `IsPlanar` and then `FitPlane` to get the normal of the outline? By these means we would avoid the code crashing on open outlines, which could actually make sense - as discussed in https://github.com/BHoM/Revit_Toolkit/issues/1008, `RoundCoordinates` could be used as a way to close the gaps, ergo it would be good if it did not crash on them.

### PR DESCRIPTION
**Pull request by LMarkowski**
### Issues addressed by this PR

Closes #2441

Removed `Normal()` query from `IElement` and `PlanarSurface` `RoundCoordinates()` methods in the engine.
Also aligned `PlanarSurface` method to the recent changes in the `Spatial_Engine` methods. Rounding coordinates of sloped surfaces is now allowed if it won't mess the planarity.
It needed moving `FitPlane(IElement)` methods to the `Spatial_Engine`.

### Test files

[Here](https://burohappold.sharepoint.com/:f:/s/BHoM/EnBoJbstUktLil5ctcKL1RMBzSu3bcJPbdlfL87qQ5YwCA?e=YWDT2k) is the test for newly added FitPlane method.
[Here](https://burohappold.sharepoint.com/:f:/s/BHoM/Es7u_pVlelBHnMn4FMLckToBg_rAo9B9mua_D35VuZAdYw?e=DZBg8N) is the test file for Spatial RoundCoordinates methods.
And [here](https://burohappold.sharepoint.com/:f:/s/BHoM/En2Ibg00v25HmGzTJBBz76cBJaFHnTPrNFJoBsnZQo7SVQ?e=l2Nzjj) is the test file for Geometrical RoundCoordinates methods. It doesn't provide with a clean validation, but maybe in this case passing the UTs would be enough.

### Changelog

- Moved `FitPlane` methods for `IElement` classes to the `Spatial_Engine`.
- Updated `RoundCoordinates` method for the `IElement2D` class in the `Spatial_Engine`.
- Updated `RoundCoordinates` method for the `PlanarSurface` class in the `Geometry_Engine`.

### Additional Comments
`FitPlane` unit test was oddly failing so it'll be for now kept on the unit test quarantine branch which is to be created once this one will be merged.

### CODE REVIEW

**Comment by LMarkowski** [comment]
@BHoMBot check unit-tests

File: Geometry_Engine/Modify/RoundCoordinates.cs
Comment (pawelbaran) [inline Geometry_Engine/Modify/RoundCoordinates.cs:294]: `normal` could be declared inside the `if` statement.

**Review by pawelbaran — CHANGES_REQUESTED** [review]
From what I can see in the tests, the results of Geometry_Engine tests for `PlanarSurface` stay unchanged, but what is alarming, the `Spatial_Engine` test results for `IElement2D` goeds from 49 correct to 48. Is that possible or did I do something wrong?

Besides that, one minor code comment.

**Comment by pawelbaran** [comment]
Ah yest, unit tests confirm my findings 😉

**Comment by LMarkowski** [comment]
> @pawelbaran: test results for IElement2D goes from 49 correct to 48.

Shoot. I forgot the #2428 is not merged yet. It should be 50 out of 50. Testing against the UT which is being updated is kinda pointless. I guess we should wait until #2428 gets merged and update from master then, or rebase this branch to 2428's. What do you think?

**Comment by pawelbaran** [comment]
Rebase to another branch sounds like something @FraserGreenroyd will not be too happy about. I would start from updating from `master`, then rebase if necessary.

**Comment by pawelbaran** [comment]
@LMarkowski I have just updated from `master`, but the test still shows 48/50 :<

**Comment by LMarkowski** [comment]
@BHoMBot check unit-tests

**Comment by rolyhudson** [comment]
I had a quick look at the grasshopper validation for `RoundCoordinates(IElement2D)` (which I think is the one referred to above) currently showing 48/50 pass. 

The validation compares the resulting coordinates of `RoundCoordinates(IElement2D)` with coordinates obtained by extracting controlpoints of the `ElementCurves`.

The validation fails for 2 `IElement2D`s whose coordinates are not rounded by `RoundCoordinates` 
They show the warning: 
`1. Rounding the coordinates of an IElement2D couldn't be achieved without losing planarity. No action has been taken.`

In the validation the un-rounded coordinates are still compared to rounded coordinates and therefore fail.

I think the issue is on the validation method not accounting for those IElement2D that cannot have coordinates rounded.
This looks like one where input of a tolerance parameter would be helpful. (As another issue under the tolerance theme https://github.com/BHoM/BHoM_Engine/issues/1773)

Meanwhile looking separately at the unit tests I observe them to all pass when running from within the UI but FitPlane and RoundCoordinates fail when tested without UI.

**Comment by LMarkowski** [comment]
@BHoMBot check unit-tests

**Comment by LMarkowski** [comment]
@BHoMBot check unit-tests

**Comment by LMarkowski** [comment]
> The validation fails for 2 `IElement2D`s whose coordinates are not rounded by `RoundCoordinates`
> They show the warning:
> `1. Rounding the coordinates of an IElement2D couldn't be achieved without losing planarity. No action has been taken.`

Good point @rolyhudson! I actually made the validation tests to pick up those cases and mark them as true. And they did a good work. The problem was a bug that I made in previous commit which stopped the code from rounding the coordinates even when it was possible. 
Now both UTs and validation scripts should be happy.

@pawelbaran, I had to improve a bit the logic behind the method. I added a few comments to make clear what's going on. Could you check if they are clear (and maybe not necessary)?

**Comment by LMarkowski** [comment]
image
And here is how the failing `FitPlane` case looks in the grasshopper. Seems to again be an issue of numerical tolerance differences between our machines and VM. Not sure how to handle it. I don't feel good about pushing it on the master. I can remove failing case or move the whole file to the [quarantined unit tests](https://github.com/BuroHappoldEngineering/ModelLaundry_Toolkit/issues/310#issuecomment-766794783).

@rolyhudson  @FraserGreenroyd fyi

**Comment by rolyhudson** [comment]
@LMarkowski I would say quarantine the unit test. 
Also it would be good to have a central store (or record) of these that includes (or references) the creation script and the resulting json dataset. 

When time permits we can dig deeper into the problem.

**Comment by LMarkowski** [comment]
@BHoMBot check unit-tests

**Review by rolyhudson — DISMISSED** [review]
LGTM

**Review by pawelbaran — CHANGES_REQUESTED** [review]
You asked for a second review @LMarkowski, so here it comes 😃 See the code comments.

I also have one thought, whether the code for `PlanarSurface` and `IElement2D` could not be unified by some means. My intuition says this does not make much sense, but did you consider that, any thoughts?

File: Spatial_Engine/Compute/FitPlane.cs
Comment (LMarkowski) [reply Spatial_Engine/Compute/FitPlane.cs:44]: I think it says so 🤔
`Always returns null for IElement0Ds.` at the end of line 41.

**Comment by LMarkowski** [comment]
@pawelbaran I think the code for `PlanarSurface` an `IElement` is as unified as it is possible.
`IElement`s don't ensure planarity which allows for simply overwriting properties, but needs additional check if formerly planar elements hasn't been de-planarized. On the other hands `PlanarSurface` has to be planar what cuts on checks, but enforce creating the returned element via `Create` method (to be sure that the result is valid).
Besides those points the code is identical.

File: Spatial_Engine/Compute/FitPlane.cs
Comment (pawelbaran) [reply Spatial_Engine/Compute/FitPlane.cs:44]: Yeah but looks a bit inconsistent to me, line 41 vs 44. The message in 44 can be read that it is possible to create a plane, if there a singular solution for point. Or am I overreacting? 🐵

**Comment by pawelbaran** [comment]
@BHoMBot check unit-tests

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check installer

**Comment by pawelbaran** [comment]
@BHoMBot check core

**Comment by pawelbaran** [comment]
@BHoMBot check serialisation

**Comment by LMarkowski** [comment]
@BHoMBot check null-handling

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check core

**Comment by pawelbaran** [comment]
@BHoMBot check installer

**Comment by pawelbaran** [comment]
@BHoMBot check serialisation

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check serialisation

**Comment by pawelbaran** [comment]
@BHoMBot check null-handling

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@FraserGreenroyd would you be OK with getting the PR merged regardless of the compliance issues?

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a CI/CD instruction. I am authorising dispensation to be granted on check ref. 2361968817

**Comment by FraserGreenroyd** [comment]
> @FraserGreenroyd would you be OK with getting the PR merged regardless of the compliance issues?

If someone wants to approve it then yes, I'm happy.

**Review by pawelbaran — APPROVED** [review]
.
..
...
....
.....

LGTM! 😃

**Pull request merged**
